### ISSUE TITLE
To Balance or not to Balance? (and how)

### ISSUE THREAD

**Issue by Blackpaw** [lizardfs/lizardfs#607]
> That is the question.
> whether to suffer the latencies of outrageous IOPS,
> or rsync against a sea of empty disk space

_Apologies to Hamlet, and humanity in general :)_

Am facing I guess a common issue - migrating data from our current DFS to lizardfs. Due tp node, disc and uptime requirements I've separated all the ZFS mirrors on the existing nodes so I could run them in parallel (I also added two extra nodes).

So far I've migrated half the VM's and results have been excellent, so ready to move the remainder and repurpose the remaining disks for lizardfs.

Trouble is, that will leave a terribly lopsided distribution on the chunkservers. Right now each chunkserver has 2 HD in RAIDZ0, there is sufficient space to store all our VM's on these, replica 3. Once that is done I'll be adding another 2 HD in RAIDZ0 as a 2nd HDD entry for the chunkservers.

As far as storage goes, probably doesn't matter - I imagine new writes will be biased to the new HDD and over time they will fill up.

IOPS however are an issue :) You can never have to many of them. I'm assuming random read/writes to existing VM's would benefit from being spread over 2 HDD entries rather than one.

So my questions are:

- Would my VM's benefit being rebalanced from a 100/0% distribution on each chunkserver to a 50/50%  spread.

If so, I see the various methods for doing so as:
1. rsync
Just rsync all the chunks from HDD1 to HDD2, then add HDD2 to the chunkserver. Hopefully it will delete the duplicate chunks in a even manner

2. replication (undergoal)
Current goal is R3, so I could just delete all data from the chunkserver and allow it to rebuild. This would involve being under goal (R2) for a extended period, which I find stressfull :)

3. replication (overgoal)
I could increase the Goal to 4, then delete the chunkserver data and let it replicate. This would take twice as long but we would never drop under R3.

1 seems a bit hinky to me, 2 & 3 go through appropriate channels with 3 being the safest (but slowest). Fortunately I am not in any hurry, I can afford to spread this over a few weekends and the volume of data (3TB) is not huge these days. Once I'm finished we will have a ridiculous amount of space available, so performance is much more important than any space wastage.

I'd be interested in peoples thoughts - is this worth doing? are there better ways of doing it?

Thanks.

**Comment by Zorlin** [comment]
This is probably a question for @4Dolio who is no stranger to this kind of juggling :)

**Comment by 4Dolio** [comment]
Use option 3. I need to think about it more before I could pick other methods

**Comment by 4Dolio** [comment]
3TB shouldn't take too long. Make sure your full tilt long duration replication is tuned such that you leave enough headroom for normal client load to not notice. As you say option 3 is safest and will give best chunk distribution.

**Comment by 4Dolio** [comment]
... I can discrete lfs chunk disks by the name Brick. Could be a single disk, a partition, a hw raid, zfs raidz0,1,5,6, thumb drives, md raid on floppies.  Who cares. Call it a brick..

so yo could just do a new raidz0 pair. Call it brick2 on each chunk server.  0-100% replication. Even ballance. One trip.

**Comment by Blackpaw** [comment]
Yes, thanks  @4Dolio, that was my thoughts too. Its all ZFS underneath, as you probably guessed originally RAIZD10 per node, which I split into two RAID RAIZ0 pairs per node - one for the old system, one for Lizard. ZFS makes these things easy and BTW, the lz4 compression has been very efficient (1.82 Ratio).

I'll turn on Goal 4 tonight, should be interesting to see how that effects performance once its finished. If its a small hit I may even leave it on - you can never have to much redundancy. Then add the 2nd RAIDZ0 pair as HDD2 (or brick2 etc, ex gluster person?) and let the chuckserver resync from scratch.

Do you think it would improve IOPS to have balanced data on the chunkservers?

**Comment by 4Dolio** [comment]
Humm.. you could be back to raidz10 or raidz01? Sound be similarly simple. no, wait, could you? If you have stripes now, you could mirror to a new one. or try one of each method and then choose..

**Comment by Blackpaw** [comment]
> If you have stripes now, you could mirror to a new one.

Sorry, not sure what you mean by that?

**Comment by 4Dolio** [comment]
You said you have raidz0 which is a stripe? Instead of a second stripe as a brick. It could mirror the current stripe.  Back to RAIZD10 you called it (zfs raid1+0 or raid0+1 whatever). Idk. Nvm. Dohh. That would be silly. You already have 3x redundancy, or 4x...

zfs compression is great. Love it.

Goal=4 would be a descent rep test to do.  That should read from 3CS and write to 1CS (to keep in mind for rep tuning. May as well keep it there if you have the capacity to burn.  You could in theory drop to any single member and still be online with goal=4.

I am curious what you are stacking on top of the LFS now...

Humm. Is that where I picked up the Bricks term? We did try gluster but choose MooseFS instead.  gluster was not cutting it but all I recall is that I wasn't thrilled with the mechanics of the thing, it kept getting overloaded and/or dropping data when we threw clients at it, maybe it didn't play well with nfs? Shrugs.

Anyway, back to the plan... I think if you simply add a new empty second raidz0 which is empty.  Rebalance replication should kick in.  It won't use any local disk-disk transfers, but it should result in a fairly well balanced randomization of chunk file distribution on both bricks.  Half added to new brick and half removed from the current brick.

Doing this gradually, one new brick at a time, will minimize background replication overhead at the expense of time, which you have it sounds like.  That would be a read from 3CS and write to 1CS replication ratio.

Adding all 4 new bricks could induce 3x~4x the replication load. Read from 3xCS write to 3xCS. But you could use that to tune down your maximum burst replication (read/write limits) tuning so it never saturates and squeezes your clients needs...

If replication tuning is still as it was in 2.6 then you set a loop cycle time, default is 300 5min.  And a max chunks read per CS per cycle. And a max chunks written per CS per cycle. If you have too large of values you just saturate your weakest component (disk, net, etc).  The over saturation is harmless in that failures to rep a given chunk will just try again later.  So you can step up your limits and HUP the master untill you stop gaining replication rate per the master graphs.  Should make clear your bottleneck.  Clients might not be happy during the overload.  But you will know your limits and can back off and better gauge how much to back off to leave plenty of overhead leftover for clients.

**Comment by Blackpaw** [comment]
> I am curious what you are stacking on top of the LFS now...

Just Windows VM (qemu/kvm)

>Anyway, back to the plan... I think if you simply add a new empty second raidz0 which is empty. Rebalance replication should kick in. It won't use any local disk-disk transfers, but it should result in a fairly well balanced randomization of chunk file distribution on both bricks. Half added to new brick and half removed from the current brick.


Ok, that is contrary to what I thought - lizard will rebalance between disks on a chunkserver?

I currently have a rep 3 goal "3 3: VNH _ _", where VNH is the first server I am going to double the disks on. It also has by far the fastest CPU and Disks, which is why its listed as needing 1 copy of everything.

nb: We also have 5 chunkservers, not 4.

**Comment by 4Dolio** [comment]
No. It wont be rebalancing the two local disks on the chunk server.  It will be rebalancing  100%+0% CS against many peers at 100%. With replication over the network.  A dev can clarify, but the CS will always try to equally fill disks. And if they are somehow more than way 5% or 10%, idk exactly. They should treat the 0% as if it there were undergoal chunks and also treat the 100% as if it were overgoal.  It should rebalance to say 55%+45% or something, which should be close enough.  I could be wrong, someone should confirm my explanation.. To be clear there is no direct disk to disk replication among local disks.

**Comment by psarna** [comment]
Right now replication (and thus rebalancing as well) occurs between chunkservers, not between disks, just like @4Dolio states.

True, rebalancing between local disks would be really useful and it's definitely implementable. The only reason why it's not done yet is that we're busy doing other things and it would require some time to implement it and test it properly :) That's it.

**Comment by Blackpaw** [comment]
Understood, Piotr, thanks for the clarification. I'm happy with the workarounds above - in no particular hurry.

One of the things I'm wary of on my part is ending up working against the product by over finessing it. I am tempted to just add the disks and let lizard do as it sees fit.

I have come to realise that for any DFS, 3 nodes is a bare minimum and even that leaves you with no wiggle room whatsoever. Taking the nodes to 5 has made life much easier and ability to sustain performance definitely seems improved.

**Comment by Blackpaw** [comment]
>If replication tuning is still as it was in 2.6 then you set a loop cycle time, default is 300 5min. And a max chunks read per CS per cycle. And a max chunks written per CS per cycle. If you have too large of values you just saturate your weakest component (disk, net, etc). 

Wee! No one uses the cluster other night so I set Goal 4, increased all the replication values by an order of magnitude and suspended all the VM's. Total bandwidth is averaging 950Mbs send/receive and peaking regularly at 1.8Gbps. Not bad for 3*1G bonds. It'll be finished in a couple of hours at this rate.

Cluster comms has a separate 1G subnet, so its coping fine.

**Comment by guestisp** [comment]
> Wee! No one uses the cluster at night so I set Goal 4, increased all the replication values by an order of magnitude and suspended all the VM's. Total bandwidth is averaging 950Mbs send/receive and peaking regularly at 1.8Gbps. Not bad for 3*1G bonds. It'll be finished in a couple of hours at this rate.

How do you get these speeds? Which bonding mode ?

**Comment by Blackpaw** [comment]
linux bond (not OVS), balance-alb. And its because there are five nodes all sending/receiving to each other simultaneously, hence all three parts of the bond get used. The aggregate bandwidth works out quite high.

iowait is also quite high ;) average 25% peaking as high as 60%. The VM's would be unusable if anyone was trying to use them.

Could probably get even higher if I disabled PERFORM_SYNC, but I can't be bothered.

**Comment by guestisp** [comment]
I'm also testing `balance-alb` and `balance-rr` but with no success.

With `balance-rr` iperf is able to saturate 2gbps (with 2 nics), with `balance-alb` it can't go faster than 1gbps

With both, inside a VMs booted from Lizard, i'm unable to reach more than 30-40MB/s (320mbit) regardless `PERFORM_SYNC` set to 1 or 0

**Comment by Blackpaw** [comment]
Well bumping replication goal up to 4 had no effect on performance, if anything it improved. I'm very very pleased. Will be rebalancing the VNH chunkserver tonight (i,e erasing the data, adding the extra disks and letting it resync).

It will take a while, but it should be interesting to see how the cluster performs once I've finished this for all of the chunkservers.

**Comment by Blackpaw** [comment]
Heh, I misread the meaning of CHUNKS_LOOP_MIN_TIME and increased rather than decreased it, now have it down to 3 seconds. Now averaging 1.5Mbps transfer for the re-population of VNH.

**Comment by guestisp** [comment]
> Heh, I misread the meaning of CHUNKS_LOOP_MIN_TIME and increased rather than decreased it, now have it down to 1 second. Averaging 1.5Mbps transfer for the re-population of VNH.

Having the same issue here. #600 

Based on official docs, higher the `CHUNKS_LOOP_MIN_TIME` value the more chunks are processes in one run. Or not ?

**Comment by Blackpaw** [comment]
Not:
>Chunks loop will check all chunks in specified time (default is 300) unless CHUNKS_LOOP_MAX_CPS will force slower execution.

It specifies the number of seconds it will take to process the chunks, so the lower the value the faster it processes them. So specifying a value 1 means it takes once second to process the chunks rather than 300 (subject to CHUNKS_LOOP_MAX_CPS)

It did take me several reads to realise that :)

**Comment by guestisp** [comment]
I think that docs about these values are not wrote properly or not clear enough (but english is not my native language, maybe I misunderstood something)

So, setting that value to 60, how many chunks are processed in 60 seconds ? You can't know how many chunks are processable in one run. 

And why lowering that value will make lizard faster ?

It makes more sense to set how long the loop should run knowing that Lizards will parse all chunks as it can up to defined seconds.

Disks are speed limited, thus allowing a longer loop execution, should allow a longer "scrub"

**Comment by guestisp** [comment]
If you have 4 millions chunks, you won't be able to check the mall in `CHUNKS_LOOP_MIN_TIME` seconds
So, how many chunks are checked in `CHUNKS_LOOP_MIN_TIME` ?

And if that is the loop time length, it should be called `CHUNKS_LOOP_**MAX**_TIME`, not MIN, as it define the max length for a chunk loop.

Really, is not clear to me.

**Comment by psarna** [comment]
>Chunks loop will check all chunks in specified time (default is 300) unless CHUNKS_LOOP_MAX_CPS will force slower execution.

Master will *try* to check all chunks in CHUNKS_LOOP_MIN_TIME seconds. If there are too many chunks to do that, it process as many as it can in given time. Hence, it's min time.
So, if there are 5 chunks in your system and CHUNKS_LOOP_MIN_TIME is 300, it will check more or less 1 chunk every 60 seconds, saving a lot of CPU cycles for other things.
If there are 1000000000 chunks in your system, it will process as many as it can. After 300 seconds it will continue checking the rest in the next loop. Hence, it's min time.

**Comment by guestisp** [comment]
So, lowering doesn't increase speed. Speed is increased only if you have a very small chunk number where chunks can be checked totally in less than CHUNKS_LOOP_MIN_TIME

In example, I have 538 chunks, i've set CHUNKS_LOOP_MIN_TIME to 1 and manually rotted a single file about 40 minutes ago.

chunk is still rotted. I would expect to be fixed in about 538 seconds (538 chunks, 1 loop per second, thus more or less 538 seconds to parse all chunks)

**Comment by psarna** [comment]
If you rotted a single file, then it needs to be discovered by a chunkserver. CHUNKS_LOOP_MIN_TIME has nothing to do with it. Chunk loop in master checks metadata - so it will check which chunks are undergoal, needs to be rebalanced, deleted, replicated, etc. It has nothing to do with file data. Chunk rot can be discovered by a chunkserver, either by periodic chunk testing or performing any I/O operation on it.

**Comment by guestisp** [comment]
Ok, now it's clear.

So, If I loose a single chunk, that missing chunk will be discovered by master in every `CHUNKS_LOOP_MIN_TIME` but a rotted file is discovered by chunkserver in every `HDD_TEST_FREQ` ?

`HDD_TEST_FREQ` will check ONE single chunk every `HDD_TEST_FREQ` seconds ?
So, with 500 chunks, I need 500*`HDD_TEST_FREQ` seconds to scan them all ?

**Comment by guestisp** [comment]
I've set `HDD_TEST_FREQ` to 1, and reloaded the chunkserver process.
Stracing that process I can see `open` every 10 seconds. restarting instead of a simple reload, fixed this and opens are every second.

Then i've tried to set `HDD_TEST_FREQ` to 2 and reloaded again. Again, no changes in frequency.
The only way to change frequency is to restart the chunkserver.

Is this ok? Isn't possible to reload chunkserver configuration without a restart?

**Comment by 4Dolio** [comment]
It is worth pointing out that this is the case for a bunch of settings.. Some are picked up during a reload and others require a restart.  Not sure to the rhyme or reason behind why some work and others done... The great thing is that it doesn't matter all that much since the cluster should easily survive the restart of any component of the cluster so if you have to resort to a restart it is relatively harmless.  Of course that might mean figuring out what you can not tune on the fly with reloads and paying closer attention to those while making a concerted effort to explore and note acceptable lower and upper limits for some parameters, so you do not have to blindly guess at things when you least want to be restarting things.

I think many of the things you might want to dynamically change with reloads do work that way.   But I have also sometimes wanted to fiddle with something but did not because it requires restarting.  Would be nice to have an official list of all tunables and whether they get reevaluated during a HUP or not.

**Comment by guestisp** [comment]
> It is worth pointing out that this is the case for a bunch of settings.. Some are picked up during a reload and others require a restart. Not sure to the rhyme or reason behind why some work and others done... 

Good to hear, so it not my mistake.

Anyway, can you tell me how `CHUNKS_LOOP_MIN_TIME` works? I'm still trying to figure it out.

And is true saying that every `HDD_TEST_FREQ` seconds chunkserver will check **ONE** chunk?
If checked chunk is rotted, replication doesn't seem to start immediatly, i'm figured this out by stracing the chunkserver process. If I understood properly, a rotted file is only deleted by chunkserver.
Then (let me know if i'm wrong), the deleted chunk will be catched by master within `CHUNKS_LOOP_MIN_TIME` that will start replication as missing, right ?

**Comment by psarna** [comment]
Yes, it's true, chunkservers test one chunk per `HDD_TEST_FREQ` seconds. After chunkserver detects a damaged chunk, it reports it immediately to master. Master acknowledges it and will order a replication during the next chunk loop.

Also, if this particular chunk becomes endangered after losing this copy and `ENDANGERED_CHUNKS_PRIORITY` is set in master config, its replication will most likely happen faster.

**Comment by 4Dolio** [comment]
Just for kicks we once calculated the chunk check mechanic would verify all our chunks in just 30 short years..  So it is not really equivalent to a raid verify or a zfs resilver.  It is more like the canary in the coal mine as an early indication of some greater underlying disk media problems.  I think we have already discusssed how to try and replicate a lfs resilver like process.

At home i use the chunk checks to just keep the drives spun up with about 1 read per minute.  At work i think we also turned checks way down or nearly off. To each his own.

**Comment by guestisp** [comment]
> Just for kicks we once calculated the chunk check mechanic would verify all our chunks in just 30 short years.. So it is not really equivalent to a raid verify or a zfs resilver. It is more like the canary in the coal mine as an early indication of some greater underlying disk media problems. I think we have already discusssed how to try and replicate a lfs resilver like process.

You are right, that's why a scrub-like feature is almost mandatory and I've asked for that in another issue.
The current "scrub" system in Lizard is almost useless with millions chunks. You'll need years to check.

One of our smallest mail server has about 5.000.000 files that means at least 5 millions chunks.
Even by setting `HDD_TEST_FREQ` to 1,  we'll need about 58 days to scrub everything (5.000.000÷3.600÷24 = 58). If these file are not actively used (we have 8 years maildirs full of archived emails), won't be checked for months. And when that file read back again, could be lost because Lizard didn't notice (for the same reason) that it's replica on the other server is missing.

And that's is only one single mail server, we have about 6 of them, plus any other server (always in the 4 millions range)
